Week8/app/src/main/java/com/example/marvelcharacters/ui/MainActivity.kt

```diff
@@ -32,25 +27,36 @@ class MainActivity : AppCompatActivity() {
         initCharacterGrid()
         viewModel.getAllCharacters().observe(this, Observer {
             characterAdapter.updateCharacters(it)
-            if (characterAdapter.itemCount == 0) {
-                getCharactersFromApi()
-            }
         })
+        updateBatchCharactersFromApi()
     }
 
-    private fun getCharactersFromApi() {
-        networkStatusChecker.performIfConnectedToInternet {
-            lifecycleScope.launch(Dispatchers.Main) {
-                val result = remoteApi.getCharacters()
-                if (result is Success) {
-                    viewModel.insert(result.data)
-                } else {
-                    toast(
-                        getString(R.string.error_message)
-                    )
-                }
+    /**
+     * Runs background API requests periodically to check for updated information
+     * */
+    private fun updateBatchCharactersFromApi() {
+        val constraints = Constraints.Builder()
+            .setRequiredNetworkType(NetworkType.CONNECTED)
+            .setRequiredNetworkType(NetworkType.NOT_ROAMING)
+            .setRequiresBatteryNotLow(true)
+            .setRequiresStorageNotLow(true)
+            .build()
+
+        val remoteApiWorker = PeriodicWorkRequestBuilder<RemoteApiWorker>(15, TimeUnit.MINUTES)
+            .setConstraints(constraints)
+            .build()
+
+        val workManager = WorkManager.getInstance(this)
+        workManager.enqueue(remoteApiWorker)
```

Comment:
And this would be yet another method. The code would be much more readable that way and every method would have just one responsibility.

Reply:
I also recommend that you move this entire code responsible for building the worker, outside of the Activity. I would probably put it into RemoteApiWorker.